### PR TITLE
Support posting messages to semgrep-app

### PR DESCRIPTION
**Pull request by underyx**

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -11,6 +11,12 @@ inputs:
   slackWebhookURL:
     description: |
       A Slack incoming webhook URL that you want findings posted to.
+  semgrepAppToken:
+    description: |
+      Access token to post results to the semgrep app.
```
Comment (nbrahms) [inline action.yml:16]: In some magical future this should be pulled out of GitHub secretes

File: entrypoint.sh [hunk 2]
```diff
@@ -4,13 +4,23 @@ set -euo pipefail
 bento_path=$(which bento)
```
Comment (nbrahms) [inline entrypoint.sh:4]: At some point we should probably just make this a Python script

File: semgrep-monitor [hunk 3]
```diff
@@ -50,11 +75,40 @@ def generate_message(notify_reason: str) -> Dict[str, Any]:
     ]
 
 
-def main():
+def report_start_to_semgrep_app():
+    if not args.semgrep_app_token:
+        return
```
Comment (nbrahms) [inline semgrep-monitor]: Some sort of error / warning logging would be useful here

File: semgrep-monitor [hunk 4]
```diff
@@ -73,13 +130,57 @@ def main():
...
-            urllib.parse.urlencode({"payload": json.dumps(payload)}).encode(),
+        response = requests.post(args.slack_url, data={"payload": json.dumps(payload)},)
+        response.raise_for_status()
+    except requests.RequestException:
+        print(f"Slack returned this error: {response.text}", file=sys.stderr)
+
+
+def report_finish_to_semgrep_app():
+    if not args.semgrep_app_token:
+        return
```
Comment (nbrahms) [inline semgrep-monitor]: Ditto here

File: semgrep-monitor [hunk 5]
```diff
@@ -73,13 +130,57 @@ def main():
...
+
+    # send findings
+    results = json.load(args.results_file)
+    payload = [
```
Comment (nbrahms) [inline semgrep-monitor]: We may want to batch this in groups of 10000 or so to keep the payload size below 10 MB.

File: semgrep-monitor [hunk 6]
```diff
@@ -73,13 +130,57 @@ def main():
...
         )
-    except urllib.error.HTTPError as e:
-        print(f"Slack returned this error: {e.read().decode()}")
+        response.raise_for_status()
+    except requests.RequestException:
+        print(f"Semgrep App returned this error: {response.text}", file=sys.stderr)
+
+    # mark as complete
+    stats = {
+        "findings": len(results)
```
Comment (nbrahms) [inline semgrep-monitor]: You can put whatever stats you want here; the API example ones were just random ones I came up with in 10 seconds of brainstorming.

File: semgrep-monitor (hunk 5 above)
Comment (nbrahms) [inline semgrep-monitor]: Except for batching, we should just make `payload = results`.

There's not currently any expectation of schema on the backend.

File: action.yml (hunk 1 above)
Comment (underyx) [reply action.yml:16]: That's the intention. In the workflow config the user has to explicitly map their github secret to our input argument. There's no way for us to access a github secret directly.

File: semgrep-monitor [hunk 7]
```diff
@@ -50,11 +75,40 @@ def generate_message(notify_reason: str) -> Dict[str, Any]:
...
 
 
-def main():
+def report_start_to_semgrep_app():
+    if not args.semgrep_app_token:
+        return
+
+    api_url = f"{args.semgrep_app_url}/api/agent/deployment/{args.deployment_id}/scan"
+    metadata = {
+        "repository": repo_name,
```
Comment (nbrahms) [inline semgrep-monitor]: You should add any other relevant metadata here as well; the backend has no schema expectation.

@dlukeomalley which metadata are you looking for?

Off the top of my head, I would think that an array of committers and authors would both be useful.

File: semgrep-monitor (hunk 3 above)
Comment (underyx) [reply semgrep-monitor]: Are you sure? I expect most users would want to run semgrep-action without reporting to semgrep app. If we add a warning, they'd get that message on every run.

**Review by nbrahms — CHANGES_REQUESTED** [review]

File: entrypoint.sh (hunk 2 above)
Comment (underyx) [reply entrypoint.sh:4]: Yep, that's my plan as well. Using click to allow configuration via environment variables for the whole thing would be really useful, too.

File: semgrep-monitor (hunk 3 above)
Comment (nbrahms) [reply semgrep-monitor]: Maybe hide it behind a `--verbose`?

File: semgrep-monitor (hunk 5 above)
Comment (nbrahms) [reply semgrep-monitor]: (See `bento.util.batched`)

File: semgrep_agent/meta.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
+import os
+
+import click
+import git
+
+from dataclasses import dataclass
+
+
+@dataclass
+class Meta:
```
Comment (nbrahms) [inline semgrep_agent/meta.py:11]: neat

File: semgrep_agent/semgrep_app.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            self.scan_id = response.json()["scan"]["id"]
+
+    def report_results(self, results):
+        if self.token is None or self.deployment_id is None or self.scan_id is None:
+            return
+
+        # report findings
+        for chunk in chunked_iter(results.findings, 10_000):
+            try:
+                response = self.session.post(
```
Comment (nbrahms) [inline semgrep_agent/semgrep_app.py:45]: 👌

File: semgrep_agent/bento.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+def scan(ctx: click.Context) -> Results:
+    event_type = ctx.obj.event_type
+    click.echo(f"== triggered by a {event_type}")
+
+    if event_type not in ALLOWED_EVENT_TYPES:
+        fail_on_unknown_event()
+
+    bento.init()
+
+    try:
```
Comment (nbrahms) [inline semgrep_agent/bento.py:103]: Can we add a timer around this to get the wall-clock time?

**Review by nbrahms — APPROVED** [review]

**Pull request merged**